Finder Indexer detect whether Memory engine is supported

Fedik opened the pull request:
Pull Request for Issue #40368 .

### Summary of Changes

Even when Finder is require MEMORY engine to work, it still possible to work around this, by editing DB manualy.
The changes here allows to support that work around.


### Testing Instructions

Pretty complex :)

Easiest way, 
change:
```php
$db->setQuery('ALTER TABLE ' . $db->quoteName('#__finder_tokens') . ' ENGINE = MEMORY');
```
to:
```php
$db->setQuery('ALTER TABLE ' . $db->quoteName('#__finder_tokens') . ' ENGINE = MEMORY111');
```

Then try to save any Article. Without this patch you will get an error "engine not supported"


### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: 
- [x] No documentation changes for docs.joomla.org needed
- [ ] Pull Request link for manual.joomla.org: 
- [x] No documentation changes for manual.joomla.org needed

Fedik committed:
Finder Indexer detect whether Memory engine is supported

claudiosoprano commented:
the actual queries are two in the indexer

$db->setQuery('ALTER TABLE ' . $db->quoteName('#__finder_tokens') . ' ENGINE = MEMORY');

$db->setQuery('ALTER TABLE ' . $db->quoteName('#__finder_tokens_aggregate') . ' ENGINE = MEMORY');

so maybe you need to change both

Fedik commented:
> the actual queries are two in the indexer

The fix wraps both query in to try/catch.
What you see in first comment here, is a simple way to test the changes without messing around with DB setup.

richard67 commented:
@claudiosoprano If you could create a copy of your site, using the same kind of database cluster but a separate database so you don't mess up your real site, you can test this PR here by checking if it fixes your issue on that copy site.

You can find here links to a custom update URL and an update package with which you can update the copy site to 4.3-dev including the changes of this PR: https://ci.joomla.org/artifacts/joomla/joomla-cms/4.3-dev/40373/downloads/64607/

When you have tested, go to the issue tracker here https://issues.joomla.org/tracker/joomla-cms/40373 and use the blue "Test this" button at the top corner, then select the right test result and finally submit.

Thanks in advance.

claudiosoprano commented:
I will test it as soon as i can (but not until tomorrow, i'm out of my office)

claudiosoprano commented:
I connect from home, applied the dev update, tested and it work, i can create/modify/delete articles without problems<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40373">issues.joomla.org/tracker/joomla-cms/40373</a>.</sub>

richard67 commented:
Well the thing is that it is not even possible to disable the MEMORY engine on a MySQL or MariaDB because it is also used internally by the database server. The limitation comes from the replication of the Galera Cluster. See the first limitation listed on this page https://mariadb.com/kb/en/mariadb-galera-cluster-known-limitations/ 

> Currently replication works only with the InnoDB storage engine.

So if you don't have such a cluster, you can only test it like described in the testing instructions, changing the SQL statement so it uses an invalid storage engine name and so causes the exception.

@HLeithner As far as I remember from a past issue, you have some experience with Galera Clusters. Do you know if there is a way to circumvent this by configuration of the replication? Would it be possible to exclude the 2 tables `#__finder_tokens` and `#__finder_tokens_aggregate` from replication? And @Hackwar would that work because the tables are only needed during indexing, or do these tables need to be replicated in a cluster because they are also used for searching?

claudiosoprano commented:
In a Master-Master replication is not possible to replicate only some tables, but in a Master-Slave setup it can be set to have replicated only some tables

https://dba.stackexchange.com/questions/122189/replicate-only-some-tables-of-mariadb-database

inside you will find your link to limitation of Galera Cluster

HLeithner commented:
I don't use the Galera Cluster, I only use Master-Slave replication

richard67 committed:
Simplify $supported variable handling

richard67 committed:
Return when exception is caught

No need to set the internal state after this, and it would be set to the wrong state in this case, so better return here.

richard67 committed:
Code style

When doing a return (or break or continue), have an empty line before that statement if there are other statements before that one in the same code bloxk.

richard67 commented:
I have tested this item :white_check_mark: successfully on 98b29d12c937f731b34dc4db7bf872c6a7019f64<br /><br />I've successfully tested this PR with the method described in the instructions, i.e. with modifying the storage name to an invalid one to cause the exception.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40373">issues.joomla.org/tracker/joomla-cms/40373</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40373">issues.joomla.org/tracker/joomla-cms/40373</a>.</sub>

richard67 commented:
@Fedik I've successfully tested this PR, and as it has 2 tests I've set it RTC.

However, there could be made a few small optimizations to the code. I have made a PR in your fork, see https://github.com/Fedik/joomla-cms/pull/10 . If you agree with my suggestions and merge my PR, I will test again this PR here, and I'm sure @claudiosoprano would also do that.

Fedik committed:
Merge pull request #10 from richard67/4.3-dev-Fedik-finder-detect-memory-support-mod-1

Improvements for CMS PR 40373

richard67 commented:
Back to pending due to code changes.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40373">issues.joomla.org/tracker/joomla-cms/40373</a>.</sub>

richard67 commented:
@claudiosoprano Could you test again? The pull request has received some code changes. You can download the changed file "Indexer.php" here and save it at the right place in folder "administrator/components/com_finder/src/Indexer": https://raw.githubusercontent.com/joomla/joomla-cms/c95e127110134cba681782fbba05b2a467e12704/administrator/components/com_finder/src/Indexer/Indexer.php . I will also test again in a bit. Thanks in advance.

richard67 commented on administrator/components/com_finder/src/Indexer/Indexer.php:
```diff
@@ -977,13 +985,19 @@ protected function toggleTables($memory)
+                // Set the tokens aggregate table to Memory.
+                $db->setQuery('ALTER TABLE ' . $db->quoteName('#__finder_tokens_aggregate') . ' ENGINE = MEMORY');
+                $db->execute();
+            } catch (\Joomla\Database\Exception\ExecutionFailureException $e) {
```
```suggestion
            } catch (\RuntimeException $e) {
```
With the PDO driver the PR was working fine, but not with the MySQLi driver. The reason is that the MySQLi driver raised a `mysqli_sql_exception`. I have checked that with some debug output and a `get_class($e)` call.

As both the Joomla db ExecutionFailureException exception and the mysqli_sql_exception are derived from RuntimeException it works with that with both drivers.

The behaviour of the different drivers may also depend on database client versions.

It's similar to the issues I had once fixed with #39391 and #35951 .

Fedik replied:
Okay, makes sense

richard67 commented:
@claudiosoprano Ipdate: Please wait a bit with testing again. It needs another change.

Fedik committed:
Update administrator/components/com_finder/src/Indexer/Indexer.php

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

richard67 commented:
@claudiosoprano Now it's ready. Could you test again? You can download the changed file "Indexer.php" here and save it at the right place in folder "administrator/components/com_finder/src/Indexer": https://raw.githubusercontent.com/joomla/joomla-cms/d92d5ff43823285802daf9bf78d313f0c52aac9d/administrator/components/com_finder/src/Indexer/Indexer.php .

richard67 commented:
I have tested this item :white_check_mark: successfully on d92d5ff43823285802daf9bf78d313f0c52aac9d<br /><br />I've successfully tested this PR with the method described in the instructions, i.e. with modifying the storage name to an invalid one to cause the exception.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40373">issues.joomla.org/tracker/joomla-cms/40373</a>.</sub>

alikon commented:
I have tested this item :white_check_mark: successfully on d92d5ff43823285802daf9bf78d313f0c52aac9d<br /><br />with test instruction only
no replication in place<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40373">issues.joomla.org/tracker/joomla-cms/40373</a>.</sub>

alikon commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40373">issues.joomla.org/tracker/joomla-cms/40373</a>.</sub>

claudiosoprano commented:
I will test it as soon as possible

obuisard commented:
> I will test it as soon as possible

All good @claudiosoprano ?

claudiosoprano commented:
Ok it works

richard67 commented:
> Ok it works, i can't find i tested this item

@claudiosoprano You can find it here in the issue tracker: https://issues.joomla.org/tracker/joomla-cms/40373 . Just use the blue "Test this" button at the top left corner. If you can't see that button you might have to login using your github account.

obuisard committed:
Merge branch '4.3-dev' into finder-detect-memory-support

obuisard committed:
Merge branch '4.3-dev' into finder-detect-memory-support

obuisard commented:
Thank you Fedir @Fedik for the PR and Richard for your involvement.

claudiosoprano commented:
I have tested this item :white_check_mark: successfully on 70621b1bdd19db902b6db349b353d5bc543d6bd5<br /><br />Ok works<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40373">issues.joomla.org/tracker/joomla-cms/40373</a>.</sub>

richard67 commented:
@claudiosoprano Thanks for your test, but this pull request has been already merged and will be included in the upcoming 4.3.2 release, so it doesn’t need anymore to be tested.

lscorcia commented:
The proposed fix is not enough.

The ALTER TABLE instruction on Galera clusters does not fail if the table is empty (as it is on a new installation), it only fails if the table contains at least a record. If the ALTER TABLE succeeds, MySQL returns an error at the first insert into the table. The detection logic should be improved.
